Update the concepts and how esc works pages (#12960)
pulumi/docs: content/docs/esc/concepts/how-esc-works.md

    weight: 1
---

This article details how Pulumi ESC works, including its architecture, supported integrations, and core functionalities. To understand some of the core ESC concepts, like *environments*, *dynamic secrets*, and *configuration-as-code*, please read through the [ESC Concepts](/docs/esc/concepts/) doc.

## Architecture

> Pulumi ESC is a hosted service provided as part of [Pulumi Cloud](/docs/pulumi-cloud/). ESC stores your secrets and configuration and proxies access to other secret stores through provider plugins.

Reviewer: again, add a note about self-hosted

Using configuration-as-code, these secrets and configuration values are composed into *environments* which are defined in YAML. ESC then makes these environments available to targets via common mechanisms like environment variables, configuration files, and directly via the ESC API and multi-language SDKs.

> {{< figure src="/docs/esc/assets/esc-octopus-diagram.png" caption="Figure: A diagram showing the architecture of Pulumi ESC.">}}

Reviewer: can you ensure we use the image consistency? In www.pulumi.com/esc - we use the other image for "How ESC works" - here we use this octopus diagram.

## Secrets and configuration sources

> By default, Pulumi ESC stores your configuration and secrets in Pulumi Cloud. However, ESC also integrates with a variety of third-party sources through an extensible *provider* plugin model. This allows teams to use their preferred providers without needing to manually copy or paste secrets across environments. The secrets will be dynamically fetched from the third-party API and integrated into your ESC environments.

Reviewer: mention of providers here feels repetitive, we talked about it in the previous paragraph

By aggregating secrets from these providers, Pulumi ESC provides a unified interface for retrieving, using, and managing secrets, regardless of their source.

> ### Supported secret providers

Reviewer: I would call the "Dynamic Providers" and bucket them into Login providers and secret providers - just like the integrations left nav suggests

* [GCP OIDC](/docs/esc/integrations/dynamic-login-credentials/gcp-login/) and [GCP Secrets Manager](/docs/esc/integrations/dynamic-secrets/gcp-secrets/)
* [HashiCorp Vault OIDC](/docs/esc/integrations/dynamic-login-credentials/vault-login/) and [Vault Secrets](/docs/esc/integrations/dynamic-secrets/vault-secrets/)
* [1Password](/docs/esc/integrations/dynamic-secrets/1password-secrets/), [Kubernetes](/docs/esc/integrations/kubernetes/), among others.

Teams can setup [OpenID Connect integration](/docs/pulumi-cloud/oidc/) in their cloud providers to allow ESC environments to pull short-lived credentials via **OIDC** for secure, time-limited access to secrets. These credentials can then be used in both [Pulumi IaC](/docs/pulumi-cloud/esc/environments/#using-with-pulumi-iac) workflows and [external CLIs](/docs/pulumi-cloud/esc/environments/#running-third-party-commands-using-pulumi-esc-secrets-and-config) like `aws`, `kubectl`, etc.

## The ESC data model

### Configuration and Secrets as Code

Pulumi ESC uses a configuration-as-code approach. ESC environments are defined in YAML, which contain named configuration values, encrypted secrets, and provider references.

> A simple static environment can be thought of as a collection of key/value pairs. They can also contain interpolated values and complex [structured configuration](/docs/esc/environments/working-with-environments/#structured-configuration). Static secrets are also defined in YAML, but the static values are encrypted.

Reviewer: ```suggestion
A simple static environment can be thought of as a collection of key/value pairs. They can also contain interpolated values and complex [structured configuration](/docs/esc/environments/working-with-environments/#structured-configuration). Static secrets are also defined in YAML and are encrypted before they are stored to ensure security. 
```

**Example:** *Defining configuration values, structured data, and encrypted secrets in YAML.*

```yaml
values:
  name: world
  salutation: hello
  greeting: ${salutation}, ${name}

  structured_data:
    active: true
    nums:
      - 1
      - 2

  myPassword:
    fn::secret:

>       ciphertext: ZXNjeAA....

Reviewer: ciphertext is the resulting output of a secret... I wonder whether we should specify the input and what happens after the use saves. Right now if they copy this text, it won't work

```

### Dynamic and Projected Values

Dynamic secrets are defined in YAML as provider references. The interpolation mechanism can also be used to *project* these any ESC values as environment variables or files.

**Example:** *Defining dynamic secrets and projecting environment variables.*

```yaml
values:
  aws:
    login:
      fn::open::aws-login:
        oidc:
          roleArn: arn:aws:iam::01234567891011:role/some-role

>           sessionName: some-session

Reviewer: ```suggestion
          sessionName: some-session
          duration: 1h
```
